Content.Server/Radio/EntitySystems/HeadsetSystem.cs

```diff
@@ -86,21 +106,128 @@ private void OnExamined(EntityUid uid, HeadsetComponent component, ExaminedEvent
+    }
+    private void UploadChannelsFromKey(HeadsetComponent src, EncryptionKeyComponent key)
+    {
+        foreach (var j in key.Channels)
+            src.Channels.Add(j);
+    }
+    private void RecalculateChannels(HeadsetComponent src)
+    {
+        src.Channels.Clear();
+        foreach (EntityUid i in src.KeyContainer.ContainedEntities)
         {
-            if (id == "Common") continue;
-
-            var proto = _protoManager.Index<RadioChannelPrototype>(id);
-            args.PushMarkup(Loc.GetString("examine-headset-channel",
-                ("color", proto.Color),
-                ("key", proto.KeyCode),
-                ("id", proto.LocalizedName),
-                ("freq", proto.Frequency)));
+            if (TryComp<EncryptionKeyComponent>(i, out var key))
+            {
+                UploadChannelsFromKey(src, key);
+            }
         }
+    }
+
+    private void OnInteractUsing(EntityUid uid, HeadsetComponent component, InteractUsingEvent args)
+    {
+        if (!component.IsKeysExtractable || !TryComp<ContainerManagerComponent>(uid, out var storage))
+        {
+            return;
+        }
+        if (TryComp<EncryptionKeyComponent>(args.Used, out var key))
+        {
+            if (component.KeySlotsAmount > component.KeyContainer.ContainedEntities.Count)
+            {
+                if (_container.TryRemoveFromContainer(args.Used) && InstallKey(component, args.Used, key))
+                {
+                    _popupSystem.PopupEntity(Loc.GetString("headset-encryption-key-successfully-installed"), uid, Filter.Entities(args.User));
+                    _audio.PlayPvs(component.KeyInsertionSound, args.Target);
+                }
+            }
+            else
+            {
+                _popupSystem.PopupEntity(Loc.GetString("headset-encryption-key-slots-already-full"), uid, Filter.Entities(args.User));
+            }
+        }
+        if (TryComp<ToolComponent>(args.Used, out var tool))
+        {
+            if (component.KeyContainer.ContainedEntities.Count > 0)
+            {
+                if (_toolSystem.UseTool(
+                args.Used, args.User, uid,
+                0f, 0f, new String[] { "Screwing" },
+                doAfterCompleteEvent: null, toolComponent: tool))
+                {
+                    var contained = new List<EntityUid>();
+                    foreach (var i in component.KeyContainer.ContainedEntities)
+                    {
+                        contained.Add(i);
+                    }
+                    foreach (var i in contained)
+                    {
+                        if (HasComp<EncryptionKeyComponent>(i))
+                        {
+                            component.KeyContainer.Remove(i);
+                        }
+                    }
+                    component.Channels.Clear();
+
+                    _popupSystem.PopupEntity(Loc.GetString("headset-encryption-keys-all-extrated"), uid, Filter.Entities(args.User));
```

Comment:
that should be "extracted", not "extrated"

Reply:
```suggestion
                    _popupSystem.PopupEntity(Loc.GetString("headset-encryption-keys-all-extracted"), uid, Filter.Entities(args.User));
```